### PR TITLE
Modify ASM implementation for persistence

### PR DESCRIPTION
**Pull request by dmzimmerman**
This requires significant refactoring, which is nearly complete. 

Still to be done (midnight, 8/10 into 8/11):

- dealing with event relations
- ensuring all entity annotations are correct, including adding them to the individual Asm classes
- equals, hashcode, toString for `Asm`

### CODE REVIEW

**Comment by dmzimmerman** [comment]
(note: review requested so you see what I'm doing, but it's not ready to be merged yet, and I won't get to do more until the morning)

**Comment by dmzimmerman** [comment]
OK... so at this point, this doesn't actually implement ASM persistence. But I have some ideas about that. In the meantime, we probably ought to merge the changes to the class hierarchy and such, if they're OK with reviewers, and then I can take care of the endpoint stuff I was doing yesterday, and _then_ come back to this.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/Main.java [hunk 1]
```diff
@@ -247,6 +254,47 @@ private void activateEndpoints() {
...
+   * 
+   * @exception IllegalStateException if we can't get the ASM. 
+   */
+  private void initializeASM() throws IllegalStateException {
+    Persistence.beginTransaction();
+    ASM = new RLAToolASM();
+    // note that this resets the ASM every launch right now
+    ASM.stepTransition(RLATransitionFunction.ONE.value()); 
+    
+// the following is some old database code from the initial persistence
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/Main.java]: We can go ahead and delete this commented code, right?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  private final ASMEvent my_event;
+  
+  /**
+   * The end state for this transition.
+   */
+  private final ASMState my_end_state;
+  
+  /**
+   * Constructs an AsmTransition with the specified start state,
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java]: s/AsmTransition/ASMTransition. May want to check other comments of renamed classes.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  
+  /**
+   * @return the end state.
+   */
+  public ASMState endState() {
+    return my_end_state;
+  }
+  
+  /**
+   * @return a String representation of this AsmTransition
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java]: Another stale comment, should probably do a case-sensitive grep for `Asm`

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/AbstractStateMachine.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    my_transition_function = the_transition_function;
+    my_initial_state = the_initial_state;
+    my_current_state = the_initial_state;
+    my_final_states = the_final_states;
+  }
+  
+  /**
+   * @return are we in the initial state?
+   */
+  public boolean initialState() {
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/AbstractStateMachine.java]: With these naming conventions, I'd expect `initialState()` and `finalState()` to return an `ASMState`, not a `boolean`.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/UIEvent.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * @version 0.0.1
+ */
+public enum UIEvent implements Event {
+  LOGIN,
+  FETCH_INITIAL_STATE_SEND,
+  FETCH_INITIAL_STATE_RECEIVE,
+  SELECT_NEXT_BALLOT,
+  UPDATE_BOARD_MEMBER,
+  UPDATE_BALLOT_MARKS,
+  UNDEFINED
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/UIEvent.java]: Now that I'm looking at it: what is this event for?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AuthenticateStateAdministrator.java [hunk 6]
```diff
@@ -55,13 +58,22 @@ public String endpointName() {
...
+    String result = "";
     
     if (!Authentication.authenticateAs(the_request, AdministratorType.STATE)) {
       status = HttpStatus.UNAUTHORIZED_401;
       result = "Authentication failed";
     }
     
+    // Take the transition triggered by this successful authentication.
+    Main.asm().stepEvent(DoSDashboardEvent.AUTHENTICATE_STATE_ADMINISTRATOR_EVENT);
+    the_response.status(status);
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AuthenticateStateAdministrator.java]: We do this both here and on line 77 below.

**Comment by dmzimmerman** [comment]
At this point, we should be able to get the persistent ASM state from an ASM, and reconstitute an ASM from a persistent ASM state. How we use that ability (in terms of transactions, how many ASMs to track and where, etc.) is still to be determined.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/Main.java (hunk 1 above)
Comment (dmzimmerman) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/Main.java]: We can; I can store it somewhere else. I just wanted to remember what I had done for when I have to do it again.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java (hunk 2 above)
Comment (dmzimmerman) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/ASMTransition.java]: I'll do a global search and replace.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/AbstractStateMachine.java (hunk 4 above)
Comment (dmzimmerman) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/AbstractStateMachine.java]: Those methods ought probably to be named something like `isInInitialState()` and `isInFinalState()`.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/asm/UIEvent.java (hunk 5 above)
Comment (dmzimmerman) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/asm/UIEvent.java]: I'm not certain... @kiniry?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AuthenticateStateAdministrator.java (hunk 6 above)
Comment (dmzimmerman) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AuthenticateStateAdministrator.java]: Got it.

**Pull request merged**